v1_1/system.ts

```diff
@@ -0,0 +1,167 @@
+            this.keyPair.publicKey,
+            this.keyPair.privateKey
+        );
+        const qrCodeTrace = QRCodeTrace.decode(qrBuf);
+        if (qrCodeTrace.info !== info) {
+            throw new Error("not the same info");
+        }
+        const tr = {sk: qrCodeTrace.trSk, r2: qrCodeTrace.trR2};
+        if (!CryptoV1_1.verifyTrace(info, tr,
+            {
+                r1: qrCodeTrace.pTrR1,
+                r2: qrCodeTrace.pTrR2,
+            })
+        ) {
+            throw new Error("trace verification fails")
+        }
+        return {tr, start, end}
+    }
+}
+
+/**
+ * Location is used by a location owner to create the two QRCodes.
+ */
+export class Location {
+    public data: ILocationData;
+    public notificationKey = randombytes_buf(32);
+
+    constructor(public healthAuthorityPubKey: Uint8Array,
+                public name: string,
+                public location: string,
+                public room: string,
+                public locationType: number) {
+        this.data = CryptoV1_1.genCode(this.infoStr());
+    }
+
+    static infoStr(name: string, location: string, room: string, locationType: number): string {
+        return [name, location, room, locationType.toString()].join("::");
+    }
+
+    infoStr(): string {
+        return Location.infoStr(this.name, this.location, this.room, this.locationType);
+    }
+
+    /**
+     * Returns the base64 encoded protobuf-message for the location owner.
+     *
+     * @param baseURL - anything - is ignored and removed afterwards.
+     */
+    getQRtrace(baseURL: string): string {
+        const qrct = new QRCodeTrace({
+            version: 2,
+            trSk: this.data.tr.sk,
+            trR2: this.data.tr.r2,
+            pTrR1: this.data.pTr.r1,
+            pTrR2: this.data.pTr.r2,
+            info: this.infoStr(),
+        });
+        const traceBuf = QRCodeTrace.encode(qrct).finish();
+        const traceEnc = crypto_box_seal(traceBuf, this.healthAuthorityPubKey);
+        return `${baseURL}#${to_base64(traceEnc)}`;
+    }
+
+    /**
+     * Returns the base64 encoded protobuf-message necessary for visitors to register.
+     *
+     * @param baseURL - anything - is ignored and removed afterwards.
+     */
+    getQRentry(baseURL: string): string {
+        const qrCodeContent = QRCodeContent.create({
+            version: 2,
+            publicKey: this.data.ent,
+            pEnt: this.data.pEnt,
+            name: this.name,
+            location: this.location,
+            room: this.room,
+            venueType: this.locationType,
+            notificationKey: this.notificationKey
+        });
+
+        const qrCodeBuf = QRCodeContent.encode(qrCodeContent).finish();
+        return `${baseURL}#${to_base64(qrCodeBuf)}`;
+    }
+}
+
+/**
+ * The user has zero or more instances of Visit in his phone.
+ */
+export class Visit {
+    private visit: IVisit;
+
+    constructor(
+        qrCodeEntry: string,
+        public entry: number,
```

Comment:
Does this make them random public fields?

Reply:
removed the `public` and retrieved `match`